include/fw_info.rst

```diff
@@ -0,0 +1,123 @@
+
+The firmware information structure has a 12-byte magic header and a verified binary layout to ensure that the format is portable and identifiable.
+It must be located at one of three offsets from the start of the image: 0x200, 0x400, or 0x800.
+The reason that the structure is not located at 0x00 is that this can be problematic in some use cases, such as when the vector table must be located at 0x00.
+
+These rules make it simple to retrieve the information by checking for each possible offset (0x200, 0x400, 0x800) if the first 12 bytes match the magic value.
+If they do, the information can be retrieved according to the definition in :c:type:`fw_info`.
+
+Information structure
+*********************
+
+The information structure contains a minimal set of information that allows other code to reason about the firmware with no other information available.
+It includes the following information:
+
+* The size of the firmware image.
+* The single, monotonically increasing version number of the image.
+* The address through which to boot into the firmware (the vector table address).
+  This address is not necessarily the start of the image.
+
+Additionally, there is information for exchanging arbitrary data:
+
+* Application Binary Interface (ABI) getter (:cpp:member:`abi_out`)
+* Pointer to ABI getter (:cpp:member:`abi_in`)
+
+ABIs
+****
+
+The firmware information structure allows for exchange of arbitrary tagged and versioned interfaces called Application Binary Interfaces (ABIs).
+
+An ABI structure is a structure consisting of a header followed by arbitrary data.
+The header consists of the following information:
+
+* ABI ID (uniquely identifies the ABI)
+* ABI version (single, monotonically increasing version number for the ABI with this ID)
+* ABI flags (32 individual bits for indicating the particulars of the ABI)
+* ABI length (length of the following data)
+
+To retrieve an ABI, a firmware image calls another firmware image's ABI getter.
+Every image must provide an ABI getter (:cpp:member:`abi_out`) that other images can use to retrieve its ABIs.
+This ABI getter is a function pointer that retrieves ABI structs (or rather pointers to ABI structs).
+
+In addition, every image can access other ABIs through a second ABI getter (:cpp:member:`abi_in`).
+This ABI getter must be provided when booting the image.
+In other words, an image should expect its :cpp:member:`abi_in` ABI getter to be filled at the time it boots, and will not touch it during booting.
+After booting, an image can call :cpp:member:`abi_in` to retrieve ABIs from the image or images that booted it without knowing where they are located.
+
+Each image can provide multiple ABIs.
+An ABI getter function takes an index, and each index from 0 to *n* must return a different ABI, given that the image provides *n* ABIs with the same ID.
```

Comment:
This is for ABIs with the same ID, right?
But where do I get the ID? It must be provided to retrieve the ABI?

Reply:
The ID is a parameter to the ABI getter function.

Reply:
Right, but how do I know what ID to use?

Reply:
They are typically defined in the header file that has the ABI definition (EXT_ABI_FUNCTION etc).